File: python/openharness/options.py
"""Typed options container for :func:`query` and :class:`OpenHarnessClient`.

Mirrors Claude Code's ``ClaudeAgentOptions``. Bundles the 12 kwargs the
entry points accept into a single frozen dataclass so IDE discoverability
is better and config objects can be built incrementally by test helpers
or higher-level wrappers.

Usage::

    from openharness import OpenHarnessClient, OpenHarnessOptions

    opts = OpenHarnessOptions(
        model="ollama/llama3",
        permission_mode="trust",
        max_turns=5,
    )
    async with OpenHarnessClient(**opts.to_kwargs()) as client:
        ...

Unpacking with ``**opts.to_kwargs()`` lets individual kwargs override
fields inline::

    async with OpenHarnessClient(
        **opts.to_kwargs(),
        model="anthropic/claude-sonnet-4-6",
    ) as client:
        ...

We intentionally do not add a magic ``options=`` kwarg to the entry
points — that would create ambiguity when both ``options`` and an
individual kwarg are set. ``to_kwargs()`` + unpacking is explicit and
plays well with ``mypy --strict``.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable, Sequence
from dataclasses import dataclass, field, fields
from typing import Any, Literal

from ._permission_server import PermissionCallback

__all__ = ["OpenHarnessOptions", "PermissionMode"]

PermissionMode = Literal["ask", "trust", "deny", "acceptEdits", "plan", "auto", "bypassPermissions"]


@dataclass(frozen=True)
class OpenHarnessOptions:
    """All configuration for :func:`query` / :class:`OpenHarnessClient` in one object.

    Field defaults match the two entry points' defaults. See their
    docstrings for per-field semantics.
    """

    model: str | None = None
    permission_mode: PermissionMode = "trust"
    allowed_tools: Sequence[str] | None = None
    disallowed_tools: Sequence[str] | None = None
    max_turns: int = 20
    system_prompt: str | None = None
    cwd: str | None = None
    env: dict[str, str] | None = None
    tools: Sequence[Callable[..., Any] | Callable[..., Awaitable[Any]]] | None = field(default=None)
    can_use_tool: PermissionCallback | None = None
    resume: str | None = None
    setting_sources: Sequence[str] | None = None

    def to_kwargs(self) -> dict[str, Any]:
        """Return the non-None fields as a kwargs dict.

        Use with ``**`` unpacking to pass to :func:`query` or
        :class:`OpenHarnessClient`. Fields whose value is ``None`` are
        omitted so callers that want to override a field to ``None``
        must do so explicitly after unpacking.
        """
        out: dict[str, Any] = {}
        for f in fields(self):
            v = getattr(self, f.name)
            if v is not None:
                out[f.name] = v
        return out

File: packages/sdk/src/options.ts
/**
 * Public option types for `query()` and `OpenHarnessClient`.
 *
 * Mirrors `python/openharness/options.py` and `query.py`. Field shape is the
 * canonical TypeScript surface; the wire is camelCase already, so no
 * conversion is needed here.
 */

import type { PermissionCallback } from "./permissions.js";
import type { ToolDefinition } from "./tools.js";

export type PermissionMode = "ask" | "trust" | "deny" | "acceptEdits" | "plan" | "auto" | "bypassPermissions";

export interface OpenHarnessOptions {
  /** Model string (e.g. `"ollama/llama3"`, `"claude-sonnet-4-6"`). */
  model?: string;
  /** Permission gate mode for the spawned CLI. Defaults to `"trust"` so headless callers don't hang on prompts. */
  permissionMode?: PermissionMode;
  /** Whitelist of tool names. Tools outside this list are blocked. */
  allowedTools?: readonly string[];
  /** Blacklist of tool names. */
  disallowedTools?: readonly string[];
  /** Maximum number of model turns before the session ends. Defaults to 20. */
  maxTurns?: number;
  /** Override the default system prompt. */
  systemPrompt?: string;
  /** Working directory for the spawned CLI. */
  cwd?: string;
  /** Environment variables for the CLI. Merged on top of `process.env`. */
  env?: Record<string, string>;
  /** Optional override for the `oh` binary path. Equivalent to `OH_BINARY=`. */
  ohBinary?: string;
  /**
   * TypeScript tools to expose to the agent for the lifetime of this call.
   * The SDK starts an in-process MCP HTTP server, registers each tool, and
   * writes an ephemeral `.oh/config.yaml` pointing the spawned `oh` CLI at
   * it. Use {@link tool} for ergonomic construction.
   */
  tools?: ToolDefinition[];
  /**
   * Permission gate. When set, the SDK starts an in-process HTTP server and
   * routes every `permissionRequest` hook through this callback. Sync and
   * async callbacks both work. Failures (throw / timeout / unrecognised
   * return) all surface as `deny` — fail-closed.
   *
   * Requires the `oh` CLI shipped in `@zhijiewang/openharness` v2.16.0+.
   */
  canUseTool?: PermissionCallback;
  /**
   * Session ID to resume. When set, the CLI replays the prior session's
   * message history before the new prompt. Capture the ID from
   * {@link OpenHarnessClient.sessionId} (or a `SessionStart` event yielded
   * by `query()`) on a previous run.
   *
   * Requires the `oh` CLI shipped in `@zhijiewang/openharness` v2.17.0+.
   */
  resume?: string;
  /**
   * Subset of config layers to merge when the CLI reads `.oh/config.yaml`.
   * Any combination of `"user"` (`~/.oh/config.yaml`), `"project"`
   * (`./.oh/config.yaml`), `"local"` (`./.oh/config.local.yaml`). When
   * omitted, all three merge in the default precedence
   * (local > project > user).
   *
   * Requires the `oh` CLI shipped in `@zhijiewang/openharness` v2.17.0+.
   */
  settingSources?: ReadonlyArray<"user" | "project" | "local">;
}

/**
 * Companion class that bundles {@link OpenHarnessOptions} into a single
 * object you can hand around (test helpers, factory functions). Mirrors
 * Python's `OpenHarnessOptions` dataclass + `to_kwargs()`.
 *
 * @example
 * ```ts
 * const opts = new OpenHarnessOptionsBundle({
 *   model: "ollama/llama3",
 *   maxTurns: 5,
 *   settingSources: ["user", "project"],
 * });
 * const client = new OpenHarnessClient(opts.toOptions());
 * ```
 *
 * Named `OpenHarnessOptionsBundle` because the interface itself is already
 * called `OpenHarnessOptions` and TypeScript merges declarations of the
 * same name in surprising ways. Both names are exported.
 */
export class OpenHarnessOptionsBundle implements OpenHarnessOptions {
  model?: string;
  permissionMode?: PermissionMode;
  allowedTools?: readonly string[];
  disallowedTools?: readonly string[];
  maxTurns?: number;
  systemPrompt?: string;
  cwd?: string;
  env?: Record<string, string>;
  ohBinary?: string;
  tools?: ToolDefinition[];
  canUseTool?: PermissionCallback;
  resume?: string;
  settingSources?: ReadonlyArray<"user" | "project" | "local">;

  constructor(init: OpenHarnessOptions = {}) {
    Object.assign(this, init);
  }

  /**
   * Return a plain `OpenHarnessOptions` object containing only the fields
   * that were explicitly set (non-`undefined`). Useful for spreading into
   * a constructor or `query()` call.
   */
  toOptions(): OpenHarnessOptions {
    const out: OpenHarnessOptions = {};
    for (const [key, value] of Object.entries(this) as Array<[keyof OpenHarnessOptions, unknown]>) {
      if (value !== undefined) {
        (out as Record<string, unknown>)[key] = value;
      }
    }
    return out;
  }
}
